File: ios/Sources/TelegramGateway.swift
import Foundation
import TDLibFramework

private func jsonString(_ object: [String: Any]) throws -> String {
    String(data: try JSONSerialization.data(withJSONObject: object), encoding: .utf8)!
}

private final class TDLibHub {
    static let shared = TDLibHub()
    private let queue = DispatchQueue(label: "dev.msgcrypt.tdlib.receive", qos: .userInitiated)
    private let lock = NSLock()
    private var sinks: [Int32: ([String: Any]) -> Void] = [:]

    private init() {
        queue.async { [weak self] in
            while let self {
                guard let pointer = td_receive(1.0) else { continue }
                let text = String(cString: pointer)
                guard let data = text.data(using: .utf8),
                      let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                      let number = object["@client_id"] as? NSNumber else { continue }
                lock.lock(); let sink = sinks[number.int32Value]; lock.unlock()
                sink?(object)
            }
        }
    }

    func register(_ id: Int32, sink: @escaping ([String: Any]) -> Void) {
        lock.lock(); sinks[id] = sink; lock.unlock()
    }
    func remove(_ id: Int32) { lock.lock(); sinks[id] = nil; lock.unlock() }
    func send(_ id: Int32, _ object: [String: Any]) throws { td_send(id, try jsonString(object)) }
}

final class TelegramGateway: MessagingGateway {
    var event: ((GatewayEvent) -> Void)?
    private let accountID: UUID
    private let root: URL
    private let clientID: Int32
    private var phone = ""
    private var titles: [Int64: String] = [:]

    init(accountID: UUID, root: URL) {
        self.accountID = accountID
        self.root = root
        clientID = td_create_client_id()
        TDLibHub.shared.register(clientID) { [weak self] in self?.receive($0) }
    }

    func connect(account: Account, phone: String) async throws {
        guard account.telegramAPIID > 0, !account.telegramAPIHash.isEmpty else {
            throw StoreError.configuration("Введите Telegram api_id и api_hash")
        }
        guard !phone.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw StoreError.configuration("Введите номер Telegram")
        }
        self.phone = phone.trimmingCharacters(in: .whitespacesAndNewlines)
        try FileManager.default.createDirectory(at: root, withIntermediateDirectories: true,
                                                attributes: [.protectionKey: FileProtectionType.completeUntilFirstUserAuthentication])
        event?(.auth(accountID: accountID, state: .connecting, detail: ""))
        try send([
            "@type": "setTdlibParameters",
            "api_id": account.telegramAPIID,
            "api_hash": account.telegramAPIHash,
            "application_version": "0.1.0",
            "database_directory": root.appendingPathComponent("db").path,
            "database_encryption_key": "",
            "device_model": "iOS MsgCrypt",
            "enable_storage_optimizer": true,
            "files_directory": root.appendingPathComponent("files").path,
            "ignore_file_names": true,
            "system_language_code": Locale.current.language.languageCode?.identifier ?? "ru",
            "system_version": ProcessInfo.processInfo.operatingSystemVersionString,
            "use_chat_info_database": true,
            "use_file_database": false,
            "use_message_database": true,
            "use_secret_chats": false,
            "use_test_dc": false
        ])
        try sendPhone()
    }

    func submitCode(_ code: String) async throws {
        guard !code.isEmpty else { throw StoreError.configuration("Введите код") }
        try send(["@type": "checkAuthenticationCode", "code": code])
    }
    func submitPassword(_ password: String) async throws {
        guard !password.isEmpty else { throw StoreError.configuration("Введите пароль") }
        try send(["@type": "checkAuthenticationPassword", "password": password])
    }
    func loadChats() async throws { try send(["@type": "loadChats", "chat_list": ["@type": "chatListMain"], "limit": 100]) }
    func loadHistory(chatID: String, limit: Int) async throws {
        guard let id = Int64(chatID) else { throw StoreError.configuration("Неверный Telegram chat id") }
        try send(["@type": "getChatHistory", "chat_id": id, "from_message_id": 0, "offset": 0,
                  "limit": min(max(limit, 1), 100), "only_local": false,
                  "@extra": "history:\(id)"])
    }
    func sendText(chatID: String, text: String) async throws -> String {
        guard let id = Int64(chatID) else { throw StoreError.configuration("Неверный Telegram chat id") }
        let requestID = UUID().uuidString
        try send(["@type": "sendMessage", "chat_id": id,
                  "input_message_content": ["@type": "inputMessageText",
                                             "text": ["@type": "formattedText", "text": text, "entities": []],
                                             "link_preview_options": ["@type": "linkPreviewOptions", "is_disabled": true],
                                             "clear_draft": true], "@extra": requestID])
        return requestID
    }
    func logOut() async throws { try send(["@type": "logOut"]) }
    func close() { try? send(["@type": "close"]); TDLibHub.shared.remove(clientID) }

    private func sendPhone() throws {
        try send(["@type": "setAuthenticationPhoneNumber", "phone_number": phone,
                  "settings": ["@type": "phoneNumberAuthenticationSettings",
                               "allow_flash_call": false, "allow_missed_call": false,
                               "is_current_phone_number": true, "allow_sms_retriever_api": false,
                               "authentication_tokens": []]])
    }
    private func send(_ value: [String: Any]) throws { try TDLibHub.shared.send(clientID, value) }

    private func receive(_ value: [String: Any]) {
        let type = value["@type"] as? String ?? ""
        if type == "updateAuthorizationState", let auth = value["authorization_state"] as? [String: Any] {
            authChanged(auth["@type"] as? String ?? "")
        } else if type == "updateNewChat", let chat = value["chat"] as? [String: Any] {
            emitChat(chat)
        } else if type == "updateChatTitle", let id = (value["chat_id"] as? NSNumber)?.int64Value {
            titles[id] = value["title"] as? String ?? String(id)
        } else if type == "updateChatLastMessage", let id = (value["chat_id"] as? NSNumber)?.int64Value {
            emitMessage(value["last_message"] as? [String: Any], fallbackChatID: id, asChatOnly: true)
        } else if type == "updateNewMessage" {
            emitMessage(value["message"] as? [String: Any], fallbackChatID: nil, asChatOnly: false)
        } else if type == "messages", let rows = value["messages"] as? [[String: Any]] {
            rows.forEach { emitMessage($0, fallbackChatID: nil, asChatOnly: false) }
        } else if type == "error" {
            event?(.error(accountID: accountID, message: value["message"] as? String ?? "Telegram error"))
        }
    }

    private func authChanged(_ type: String) {
        switch type {
        case "authorizationStateWaitPhoneNumber": try? sendPhone(); event?(.auth(accountID: accountID, state: .waitingPhone, detail: ""))
        case "authorizationStateWaitCode": event?(.auth(accountID: accountID, state: .waitingCode, detail: ""))
        case "authorizationStateWaitPassword": event?(.auth(accountID: accountID, state: .waitingPassword, detail: ""))
        case "authorizationStateReady": event?(.auth(accountID: accountID, state: .ready, detail: phone))
        case "authorizationStateClosing": event?(.auth(accountID: accountID, state: .connecting, detail: ""))
        case "authorizationStateClosed": event?(.auth(accountID: accountID, state: .loggedOut, detail: ""))
        default: break
        }
    }

    private func emitChat(_ chat: [String: Any]) {
        guard let id = (chat["id"] as? NSNumber)?.int64Value else { return }
        let title = chat["title"] as? String ?? String(id)
        titles[id] = title
        let last = chat["last_message"] as? [String: Any]
        event?(.chat(Chat(accountID: accountID, remoteID: String(id), title: title,
                          lastText: text(last) ?? "", lastAt: date(last),
                          unread: (chat["unread_count"] as? NSNumber)?.intValue ?? 0, secureState: .none)))
    }

    private func emitMessage(_ message: [String: Any]?, fallbackChatID: Int64?, asChatOnly: Bool) {
        guard let message, let id = (message["chat_id"] as? NSNumber)?.int64Value ?? fallbackChatID,
              let body = text(message) else { return }
        let title = titles[id] ?? String(id)
        if asChatOnly {
            event?(.chat(Chat(accountID: accountID, remoteID: String(id), title: title, lastText: body,
                              lastAt: date(message), unread: 0, secureState: .none)))
        } else {
            event?(.text(accountID: accountID, chatID: String(id), chatTitle: title,
                         providerID: String((message["id"] as? NSNumber)?.int64Value ?? 0), text: body,
                         timestamp: date(message), outgoing: message["is_outgoing"] as? Bool ?? false))
        }
    }

    private func text(_ message: [String: Any]?) -> String? {
        guard let content = message?["content"] as? [String: Any], content["@type"] as? String == "messageText",
              let formatted = content["text"] as? [String: Any] else { return nil }
        return formatted["text"] as? String
    }
    private func date(_ message: [String: Any]?) -> Date {
        Date(timeIntervalSince1970: (message?["date"] as? NSNumber)?.doubleValue ?? 0)
    }
}

File: android/app/src/main/java/dev/msgcrypt/app/provider/TelegramGateway.java
package dev.msgcrypt.app.provider;

import android.os.Build;

import dev.msgcrypt.app.model.Account;
import dev.msgcrypt.app.model.AuthState;
import dev.msgcrypt.app.model.Chat;
import dev.msgcrypt.app.model.SecureState;

import org.drinkless.tdlib.Client;
import org.drinkless.tdlib.TdApi;

import java.io.File;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;
import java.util.concurrent.ConcurrentHashMap;

/** Telegram transport backed by the prebuilt Android TDLib JNI AAR. */
public final class TelegramGateway extends UnavailableGateway {
    private final File root;
    private final int apiId;
    private final String apiHash;
    private final Map<Long, String> titles = new ConcurrentHashMap<>();

    private volatile Client client;
    private volatile String accountId;
    private volatile String phoneNumber;
    private volatile boolean phoneSubmitted;
    private volatile boolean loggingOut;

    public TelegramGateway(File root, int apiId, String apiHash) {
        this.root = root;
        this.apiId = apiId;
        this.apiHash = apiHash == null ? "" : apiHash;
    }

    @Override
    public synchronized CompletableFuture<Void> connect(Account account, String phone) {
        if (apiId <= 0 || apiHash.isBlank()) return failed("Введите Telegram api_id и api_hash");
        if (phone == null || phone.isBlank()) return failed("Введите номер Telegram в международном формате");
        if (client != null) return CompletableFuture.completedFuture(null);

        accountId = account.id;
        phoneNumber = phone.trim();
        loggingOut = false;
        phoneSubmitted = false;
        try {
            requireDirectory(root);
            requireDirectory(new File(root, "db"));
            requireDirectory(new File(root, "files"));
            Client.setLogMessageHandler(1, (level, message) -> { });
            emitState(AuthState.CONNECTING, "");
            client = Client.create(this::onUpdate,
                    error -> report("Ошибка обработчика Telegram", error),
                    error -> report("Ошибка Telegram", error));
            return CompletableFuture.completedFuture(null);
        } catch (Throwable error) {
            client = null;
            report("Не удалось запустить TDLib", error);
            return failed("Не удалось загрузить Android TDLib: " + message(error));
        }
    }

    @Override
    public CompletableFuture<Void> submitCode(String accountId, String code) {
        if (code == null || code.isBlank()) return failed("Введите код Telegram");
        return send(new TdApi.CheckAuthenticationCode(code.trim())).thenApply(ok -> null);
    }

    @Override
    public CompletableFuture<Void> submitPassword(String accountId, String password) {
        if (password == null || password.isBlank()) return failed("Введите пароль 2FA");
        return send(new TdApi.CheckAuthenticationPassword(password)).thenApply(ok -> null);
    }

    private void onUpdate(TdApi.Object object) {
        if (object instanceof TdApi.UpdateAuthorizationState) {
            handleAuthorizationState(((TdApi.UpdateAuthorizationState) object).authorizationState);
        } else if (object instanceof TdApi.UpdateNewChat) {
            upsertChat(((TdApi.UpdateNewChat) object).chat);
        } else if (object instanceof TdApi.UpdateChatTitle) {
            TdApi.UpdateChatTitle update = (TdApi.UpdateChatTitle) object;
            titles.put(update.chatId, update.title);
        } else if (object instanceof TdApi.UpdateChatLastMessage) {
            TdApi.UpdateChatLastMessage update = (TdApi.UpdateChatLastMessage) object;
            upsertLast(update.chatId, update.lastMessage);
        } else if (object instanceof TdApi.UpdateNewMessage) {
            emitMessage(((TdApi.UpdateNewMessage) object).message);
        }
    }

    private synchronized void handleAuthorizationState(TdApi.AuthorizationState state) {
        authorizationChanged(state);
    }

    private void authorizationChanged(TdApi.AuthorizationState state) {
        if (state instanceof TdApi.AuthorizationStateWaitTdlibParameters) {
            TdApi.SetTdlibParameters parameters = new TdApi.SetTdlibParameters(
                    false,
                    new File(root, "db").getAbsolutePath(),
                    new File(root, "files").getAbsolutePath(),
                    accountId.getBytes(StandardCharsets.UTF_8),
                    true,
                    true,
                    true,
                    false,
                    apiId,
                    apiHash,
                    "ru",
                    Build.MANUFACTURER + " " + Build.MODEL,
                    "Android " + Build.VERSION.RELEASE,
                    "MsgCrypt 0.1.0");
            authRequest(parameters, "Не удалось настроить TDLib");
        } else if (state instanceof TdApi.AuthorizationStateWaitPhoneNumber) {
            emitState(AuthState.WAITING_PHONE, "");
            if (!phoneSubmitted) {
                phoneSubmitted = true;
                authRequest(new TdApi.SetAuthenticationPhoneNumber(phoneNumber, null),
                        "Telegram отклонил номер телефона");
            }
        } else if (state instanceof TdApi.AuthorizationStateWaitCode) {
            emitState(AuthState.WAITING_CODE, "");
        } else if (state instanceof TdApi.AuthorizationStateWaitPassword) {
            TdApi.AuthorizationStateWaitPassword password = (TdApi.AuthorizationStateWaitPassword) state;
            emitState(AuthState.WAITING_PASSWORD, password.passwordHint == null ? "" : password.passwordHint);
        } else if (state instanceof TdApi.AuthorizationStateReady) {
            emitState(AuthState.READY, phoneNumber);
        } else if (state instanceof TdApi.AuthorizationStateLoggingOut
                || state instanceof TdApi.AuthorizationStateClosing) {
            emitState(AuthState.CONNECTING, "Закрытие");
        } else if (state instanceof TdApi.AuthorizationStateClosed) {
            client = null;
            if (loggingOut) emitState(AuthState.LOGGED_OUT, "");
        }
    }

    private void authRequest(TdApi.Function<TdApi.Ok> request, String failureMessage) {
        send(request).whenComplete((ok, error) -> {
            if (error != null) report(failureMessage, unwrap(error));
        });
    }

    @Override
    public CompletableFuture<Void> loadChats(String accountId) {
        return send(new TdApi.LoadChats(new TdApi.ChatListMain(), 100)).handle((ok, error) -> {
            if (error != null && !message(unwrap(error)).contains("404")) {
                throw new CompletionException(unwrap(error));
            }
            return null;
        });
    }

    @Override
    public CompletableFuture<Void> loadHistory(String accountId, String chatId, int limit) {
        long id;
        try {
            id = Long.parseLong(chatId);
        } catch (NumberFormatException error) {
            return failed("Некорректный chat id");
        }
        int safe = Math.max(1, Math.min(100, limit));
        return send(new TdApi.GetChatHistory(id, 0, 0, safe, false)).thenAccept(result -> {
            if (result.messages != null) {
                for (TdApi.Message message : result.messages) emitMessage(message);
            }
        });
    }

    @Override
    public CompletableFuture<String> sendText(String accountId, String chatId, String text) {
        long id;
        try {
            id = Long.parseLong(chatId);
        } catch (NumberFormatException error) {
            return failed("Некорректный chat id");
        }
        TdApi.InputMessageText content = new TdApi.InputMessageText(
                new TdApi.FormattedText(text, new TdApi.TextEntity[0]), null, true);
        TdApi.SendMessage request = new TdApi.SendMessage(id, null, null, null, null, content);
        return send(request).thenApply(message -> Long.toString(message.id));
    }

    @Override
    public CompletableFuture<Void> logOut(String accountId) {
        loggingOut = true;
        return send(new TdApi.LogOut()).thenApply(ok -> null);
    }

    @SuppressWarnings("unchecked")
    private <T extends TdApi.Object> CompletableFuture<T> send(TdApi.Function<T> request) {
        Client value = client;
        if (value == null) return failed("Telegram не подключён");
        CompletableFuture<T> future = new CompletableFuture<>();
        try {
            value.send(request, object -> {
                if (object instanceof TdApi.Error) {
                    TdApi.Error error = (TdApi.Error) object;
                    future.completeExceptionally(new IllegalStateException(
                            "Telegram " + error.code + ": " + error.message));
                } else {
                    future.complete((T) object);
                }
            }, future::completeExceptionally);
        } catch (Throwable error) {
            future.completeExceptionally(error);
        }
        return future;
    }

    private void upsertChat(TdApi.Chat chat) {
        if (chat == null) return;
        titles.put(chat.id, chat.title);
        String last = text(chat.lastMessage);
        long at = chat.lastMessage == null ? 0 : chat.lastMessage.date;
        GatewayListener target = listener;
        if (target != null) {
            target.onChat(new Chat(accountId, Long.toString(chat.id), chat.title,
                    last, at, chat.unreadCount, SecureState.NONE));
        }
    }

    private void upsertLast(long chatId, TdApi.Message message) {
        GatewayListener target = listener;
        if (target != null) {
            target.onChat(new Chat(accountId, Long.toString(chatId),
                    titles.getOrDefault(chatId, Long.toString(chatId)), text(message),
                    message == null ? 0 : message.date, 0, SecureState.NONE));
        }
    }

    private void emitMessage(TdApi.Message message) {
        if (message == null || !(message.content instanceof TdApi.MessageText)) return;
        TdApi.FormattedText formatted = ((TdApi.MessageText) message.content).text;
        if (formatted == null || formatted.text == null) return;
        GatewayListener target = listener;
        if (target != null) {
            target.onText(accountId, Long.toString(message.chatId),
                    titles.getOrDefault(message.chatId, Long.toString(message.chatId)),
                    Long.toString(message.id), formatted.text, message.date, message.isOutgoing);
        }
    }

    private static String text(TdApi.Message message) {
        if (message == null || !(message.content instanceof TdApi.MessageText)) return "";
        TdApi.FormattedText value = ((TdApi.MessageText) message.content).text;
        return value == null || value.text == null ? "" : value.text;
    }

    private void emitState(AuthState state, String detail) {
        GatewayListener target = listener;
        if (target != null) target.onAuthState(accountId, state, detail);
    }

    private void report(String message, Throwable error) {
        GatewayListener target = listener;
        if (target != null) target.onError(accountId, message, error);
    }

    private static void requireDirectory(File directory) {
        if (!directory.isDirectory() && !directory.mkdirs()) {
            throw new IllegalStateException("Не удалось создать каталог " + directory.getName());
        }
    }

    private static Throwable unwrap(Throwable error) {
        while ((error instanceof CompletionException || error instanceof java.util.concurrent.ExecutionException)
                && error.getCause() != null) {
            error = error.getCause();
        }
        return error;
    }

    private static String message(Throwable error) {
        return error.getMessage() == null ? error.toString() : error.getMessage();
    }

    @Override
    public synchronized void close() {
        listener = null;
        Client value = client;
        client = null;
        if (value != null) {
            try {
                value.send(new TdApi.Close(), ignored -> { }, ignored -> { });
            } catch (Throwable ignored) {
                // The native client may already be closed.
            }
        }
    }
}
